### PR TITLE
Add @version attribute

### PR DESCRIPTION
**Pull request by phirasit**
This is another basic information attribute attached to every record.

- will be 1 after a record is created.
- will increase by 1 after any kind of updates.
- will also increase by 1 after graph structure related to that record being altered (e.g. add/remove incident edges).

Note: version will be increased just 1 until `txn.commit()` is called.

### CODE REVIEW

File: include/nogdb_types.h [hunk 1]
```diff
@@ -247,6 +263,8 @@ namespace nogdb {
 
         uint32_t getDepth() const;
 
+        const Record &updateVersion(const Txn &txn) const;
```
Comment (kasidej) [inline include/nogdb_types.h]: Should this function is private (or protected)?

File: src/sql.hpp [hunk 2]
```diff
@@ -130,6 +130,9 @@ namespace nogdb {
             nogdb::Record toBaseRecord() const;
 
         private:
+
+            friend nogdb::Record;
```
Comment (kasidej) [inline src/sql.hpp]: Why `nogdb::sql_parser::Record` class is friends of `nogdb::Record`? I think one way relationship is enough.

**Comment by phirasit** [comment]
Both are fixed right now as you suggested.

File: include/nogdb_types.h [hunk 3]
```diff
@@ -197,12 +200,15 @@ namespace nogdb {
 
     class Record {
     public:
+
+        using RecordPropertyType = std::map<std::string, Bytes>;
```
Comment (pj4dev) [inline include/nogdb_types.h]: Why in `class Record`? 

For `RecordPropertyType`, I think it made me a little bit confused at the first time I saw. `RecordProperty` may be a better name (when it is placed outside `class Record` so the developers can use it as `nogdb::RecordProperty`).

File: src/base_txn.cpp [hunk 4]
```diff
@@ -218,9 +218,9 @@ namespace nogdb {
                         WriteLock<boost::shared_mutex> _(*(ctx.dbInfoMutex));
                         (*ctx.dbInfo) = dbInfo;
                     }
-                    // allow the next txns to see the latest version and updates
-                    ctx.dbTxnStat->fetchAddMaxVersionId();
                 }
+                // allow the next txns to see the latest version and updates
+                ctx.dbTxnStat->fetchAddMaxVersionId();
```
Comment (pj4dev) [inline src/base_txn.cpp]: Why need to move this?

Comment (phirasit) [reply src/base_txn.cpp]: So, `txn version` will get increased even though the graph has not been changed. `txn version` plays an important role in record version updates.

Comment (pj4dev) [reply src/base_txn.cpp]: Good point. don't forget to fix the failure test according to this change.

File: include/nogdb_types.h (hunk 3 above)
Comment (phirasit) [reply include/nogdb_types.h]: Change the name to `DataType` for less confusion.

**Comment by phirasit** [comment]
Both points have been taken care of.

File: include/nogdb_types.h [hunk 5]
```diff
@@ -197,12 +200,15 @@ namespace nogdb {
 
     class Record {
     public:
+
+        using DataType = std::map<std::string, Bytes>;
```
Comment (pj4dev) [inline include/nogdb_types.h]: Is this a mapping between a property name and its associated data value in 'Bytes'? 'DataType' would be thought of as data type such as int, double, etc.

**Comment by phirasit** [comment]
I hope this is not confusing anymore.

File: include/nogdb_types.h [hunk 6]
```diff
@@ -197,12 +200,15 @@ namespace nogdb {
 
     class Record {
     public:
+
+        using StringToBytesMap = std::map<std::string, Bytes>;
```
Comment (pj4dev) [inline include/nogdb_types.h]: How about 'PropertyData'? having 'nogdb::Record::PropertyData' may be nice for developers. 'StringToBytesMap' is quite specific and it doesn't mean anything. Let me know what you think.

Comment (phirasit) [reply include/nogdb_types.h]: I think 'StringToBytesMap' gives a clear definition of what type it really is and how to use it. With 'ProptertyData', developers have to look up to its real definition.

**Comment by phirasit** [comment]
Txn test case is fixed.

File: include/nogdb_types.h (hunk 6 above)
Comment (phirasit) [reply include/nogdb_types.h]: 'PropertyToBytesMap' sounds great. I probably go with it.

**Comment by phirasit** [comment]
Change 'StringToBytesMap' to 'PropertyToBytesMap'

**Pull request merged**